src/jit/flowgraph.cpp

```diff
@@ -9751,123 +9751,162 @@ void                Compiler::fgCompactBlocks(BasicBlock* block, BasicBlock* bNe
-    GenTreePtr blkFirst     = block->firstStmt();
-    GenTreePtr bNextFirst   = bNext->firstStmt();
 
-    // Does the second have any phis?
-    if (bNextFirst != NULL && bNextFirst != bNextNonPhi1)
+    assert((block->IsLIR() == bNext->IsLIR()) || (block->IsLIR() && bNext->isEmpty()) || (block->isEmpty() && bNext->IsLIR()));
+    if (block->IsLIR() || bNext->IsLIR())
     {
-        GenTreePtr bNextLast = bNextFirst->gtPrev;
-        assert(bNextLast->gtNext == NULL);
+        LIR::Range blockRange = LIR::AsRange(block);
+        LIR::Range nextRange = LIR::AsRange(bNext);
 
-        // Does "blk" have phis?
-        if (blkNonPhi1 != blkFirst)
-        {
-            // Yes, has phis.
-            // Insert after the last phi of "block."
-            // First, bNextPhis after last phi of block.
-            GenTreePtr blkLastPhi;
-            if (blkNonPhi1 != NULL)
-            {
-                blkLastPhi = blkNonPhi1->gtPrev;
-            }
-            else
-            {
-                blkLastPhi = blkFirst->gtPrev;
-            }
+        GenTree* blockBegin = blockRange.Begin();
+        GenTree* blockFirstNonPhi = blockRange.FirstNonPhiNode();
 
-            blkLastPhi->gtNext = bNextFirst;
-            bNextFirst->gtPrev = blkLastPhi;
+        GenTree* nextBegin = nextRange.Begin();
+        GenTree* nextFirstNonPhi = nextRange.FirstNonPhiNode();
 
-            // Now, rest of "block" after last phi of "bNext".
-            GenTreePtr bNextLastPhi = NULL;
-            if (bNextNonPhi1 != NULL)
-            {
-                bNextLastPhi = bNextNonPhi1->gtPrev;
-            }
-            else
-            {
-                bNextLastPhi = bNextFirst->gtPrev;
-            }
+        GenTree* insertionPoint = blockFirstNonPhi != blockRange.End() ? blockFirstNonPhi : blockRange.EndExclusive();
 
-            bNextLastPhi->gtNext = blkNonPhi1;
-            if (blkNonPhi1 != NULL)
-            {
-                blkNonPhi1->gtPrev = bNextLastPhi;
-            }
-            else
-            {
-                // block has no non phis, so make the last statement be the last added phi.
-                blkFirst->gtPrev = bNextLastPhi;
-            }
+        // Does the next block have any phis?
+        if (nextBegin != nextFirstNonPhi)
+        {
+            LIR::Range nextPhis = LIR::AsRange(nextBegin, nextFirstNonPhi != nullptr ? nextFirstNonPhi : nextRange.EndExclusive());
+            nextRange.Remove(nextPhis);
 
-            // Now update the bbTreeList of "bNext".
-            bNext->bbTreeList = bNextNonPhi1;
-            if (bNextNonPhi1 != NULL)
-            {
-                bNextNonPhi1->gtPrev = bNextLast;
-            }
+            blockRange.InsertAfter(nextPhis, insertionPoint);
+            insertionPoint = nextPhis.EndExclusive();
         }
-        else
+
+        // Does the block have any other code?
+        if (nextFirstNonPhi != nextRange.End())
         {
```

Comment:
btw, it would be helpful if LIR::Range::FirstNonPhiNode() specified its return values in its header comment.

---
In reply to: [73615482](https://github.com/pgavlin/coreclr/pull/18#discussion_r73615482) [](ancestors = 73615482,73598299)